In [1]:
#import Necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

In [2]:
#uploading datasets.
match = pd.read_csv('D:\Campusx\Ipl win match probablity\matches.csv')
delivery = pd.read_csv('D:\Campusx\Ipl win match probablity\deliveries.csv')

In [3]:
match.head(5)

id    Season       city        date                        team1  \
0   1  IPL-2017  Hyderabad  05-04-2017          Sunrisers Hyderabad   
1   2  IPL-2017       Pune  06-04-2017               Mumbai Indians   
2   3  IPL-2017     Rajkot  07-04-2017                Gujarat Lions   
3   4  IPL-2017     Indore  08-04-2017       Rising Pune Supergiant   
4   5  IPL-2017  Bangalore  08-04-2017  Royal Challengers Bangalore   

                         team2                  toss_winner toss_decision  \
0  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
1       Rising Pune Supergiant       Rising Pune Supergiant         field   
2        Kolkata Knight Riders        Kolkata Knight Riders         field   
3              Kings XI Punjab              Kings XI Punjab         field   
4             Delhi Daredevils  Royal Challengers Bangalore           bat   

   result  dl_applied                       winner  win_by_runs  \
0  normal           0          Sunrisers Hyderabad           35   
1  normal           0       Rising Pune Supergiant            0   
2  normal           0        Kolkata Knight Riders            0   
3  normal           0              Kings XI Punjab            0   
4  normal           0  Royal Challengers Bangalore           15   

   win_by_wickets player_of_match                                      venue  \
0               0    Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
1               7       SPD Smith    Maharashtra Cricket Association Stadium   
2              10         CA Lynn     Saurashtra Cricket Association Stadium   
3               6      GJ Maxwell                     Holkar Cricket Stadium   
4               0       KM Jadhav                      M Chinnaswamy Stadium   

          umpire1        umpire2 umpire3  
0     AY Dandekar       NJ Llong     NaN  
1  A Nand Kishore         S Ravi     NaN  
2     Nitin Menon      CK Nandan     NaN  
3    AK Chaudhary  C Shamshuddin     NaN  
4             NaN            NaN     NaN

In [74]:
match.shape

(756, 18)

In Match Dataset there is total 756 rows and 18 columns which represent match BY match winning details from 2008 to 2019.

In [75]:
match.columns

Index(['id', 'Season', 'city', 'date', 'team1', 'team2', 'toss_winner',
       'toss_decision', 'result', 'dl_applied', 'winner', 'win_by_runs',
       'win_by_wickets', 'player_of_match', 'venue', 'umpire1', 'umpire2',
       'umpire3'],
      dtype='object')

In [76]:
delivery.head(5)

match_id  inning         batting_team                 bowling_team  over  \
0         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
1         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
2         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
3         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
4         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   

   ball    batsman non_striker    bowler  is_super_over  ...  bye_runs  \
0     1  DA Warner    S Dhawan  TS Mills              0  ...         0   
1     2  DA Warner    S Dhawan  TS Mills              0  ...         0   
2     3  DA Warner    S Dhawan  TS Mills              0  ...         0   
3     4  DA Warner    S Dhawan  TS Mills              0  ...         0   
4     5  DA Warner    S Dhawan  TS Mills              0  ...         0   

   legbye_runs  noball_runs  penalty_runs  batsman_runs  extra_runs  \
0            0            0             0             0           0   
1            0            0             0             0           0   
2            0            0             0             4           0   
3            0            0             0             0           0   
4            0            0             0             0           2   

   total_runs  player_dismissed dismissal_kind fielder  
0           0               NaN            NaN     NaN  
1           0               NaN            NaN     NaN  
2           4               NaN            NaN     NaN  
3           0               NaN            NaN     NaN  
4           2               NaN            NaN     NaN  

[5 rows x 21 columns]

In [77]:
delivery.shape

(179078, 21)

In [78]:
delivery.columns

Index(['match_id', 'inning', 'batting_team', 'bowling_team', 'over', 'ball',
       'batsman', 'non_striker', 'bowler', 'is_super_over', 'wide_runs',
       'bye_runs', 'legbye_runs', 'noball_runs', 'penalty_runs',
       'batsman_runs', 'extra_runs', 'total_runs', 'player_dismissed',
       'dismissal_kind', 'fielder'],
      dtype='object')

In Match Dataset there is total 179078 rows and 21 columns which represent ball by ball details from 2008 to 2019.

In [79]:
#applying groupby to find total runs in first inning and total runs in 2nd inning
total_score_df=delivery.groupby(['match_id','inning']).sum()['total_runs'].reset_index()

In [80]:
total_score_df

match_id  inning  total_runs
0            1       1         207
1            1       2         172
2            2       1         184
3            2       2         187
4            3       1         183
...        ...     ...         ...
1523     11413       2         170
1524     11414       1         155
1525     11414       2         162
1526     11415       1         152
1527     11415       2         157

[1528 rows x 3 columns]

In [81]:
#Fatching Total runs in 1st innings
total_score_df=total_score_df[total_score_df['inning']==1]

In [82]:
match_df=match.merge(total_score_df[['match_id','total_runs']],left_on='id',right_on='match_id')

In [83]:
match_df.head(5)

id    Season       city        date                        team1  \
0   1  IPL-2017  Hyderabad  05-04-2017          Sunrisers Hyderabad   
1   2  IPL-2017       Pune  06-04-2017               Mumbai Indians   
2   3  IPL-2017     Rajkot  07-04-2017                Gujarat Lions   
3   4  IPL-2017     Indore  08-04-2017       Rising Pune Supergiant   
4   5  IPL-2017  Bangalore  08-04-2017  Royal Challengers Bangalore   

                         team2                  toss_winner toss_decision  \
0  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
1       Rising Pune Supergiant       Rising Pune Supergiant         field   
2        Kolkata Knight Riders        Kolkata Knight Riders         field   
3              Kings XI Punjab              Kings XI Punjab         field   
4             Delhi Daredevils  Royal Challengers Bangalore           bat   

   result  dl_applied                       winner  win_by_runs  \
0  normal           0          Sunrisers Hyderabad           35   
1  normal           0       Rising Pune Supergiant            0   
2  normal           0        Kolkata Knight Riders            0   
3  normal           0              Kings XI Punjab            0   
4  normal           0  Royal Challengers Bangalore           15   

   win_by_wickets player_of_match                                      venue  \
0               0    Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
1               7       SPD Smith    Maharashtra Cricket Association Stadium   
2              10         CA Lynn     Saurashtra Cricket Association Stadium   
3               6      GJ Maxwell                     Holkar Cricket Stadium   
4               0       KM Jadhav                      M Chinnaswamy Stadium   

          umpire1        umpire2 umpire3  match_id  total_runs  
0     AY Dandekar       NJ Llong     NaN         1         207  
1  A Nand Kishore         S Ravi     NaN         2         184  
2     Nitin Menon      CK Nandan     NaN         3         183  
3    AK Chaudhary  C Shamshuddin     NaN         4         163  
4             NaN            NaN     NaN         5         157

In [84]:
match_df['team1'].unique()

array(['Sunrisers Hyderabad', 'Mumbai Indians', 'Gujarat Lions',
       'Rising Pune Supergiant', 'Royal Challengers Bangalore',
       'Kolkata Knight Riders', 'Delhi Daredevils', 'Kings XI Punjab',
       'Chennai Super Kings', 'Rajasthan Royals', 'Deccan Chargers',
       'Kochi Tuskers Kerala', 'Pune Warriors', 'Rising Pune Supergiants',
       'Delhi Capitals'], dtype=object)

In [85]:
teams=['Sunrisers Hyderabad',
       'Mumbai Indians',
       'Royal Challengers Bangalore',
       'Kolkata Knight Riders',
       'Kings XI Punjab',
       'Chennai Super Kings',
       'Rajasthan Royals',
       'Delhi Capitals'
      ]

In [86]:
match_df['team1']=match_df['team1'].str.replace('Delhi Daredevils','Delhi Capitals')
match_df['team2']=match_df['team2'].str.replace('Delhi Daredevils','Delhi Capitals')
match_df['team1']=match_df['team1'].str.replace('Deccan Chargers','Sunrisers Hyderabad')
match_df['team2']=match_df['team2'].str.replace('Deccan Chargers','Sunrisers Hyderabad')

In [87]:
match_df=match_df[match_df['team1'].isin(teams)]
match_df=match_df[match_df['team2'].isin(teams)]

In [88]:
match_df.shape

(641, 20)

In [89]:
match_df['dl_applied'].value_counts()

0    626
1     15
Name: dl_applied, dtype: int64

In [90]:
match_df=match_df[match_df['dl_applied']==0]

In [91]:
match_df = match_df[['match_id','city','winner','total_runs']]

In [92]:
match_df.head(3)

match_id       city                       winner  total_runs
0         1  Hyderabad          Sunrisers Hyderabad         207
4         5  Bangalore  Royal Challengers Bangalore         157
6         7     Mumbai               Mumbai Indians         178

In [93]:
delivery_df=match_df.merge(delivery,on='match_id')

In [94]:
delivery_df.head(3)

match_id       city               winner  total_runs_x  inning  \
0         1  Hyderabad  Sunrisers Hyderabad           207       1   
1         1  Hyderabad  Sunrisers Hyderabad           207       1   
2         1  Hyderabad  Sunrisers Hyderabad           207       1   

          batting_team                 bowling_team  over  ball    batsman  \
0  Sunrisers Hyderabad  Royal Challengers Bangalore     1     1  DA Warner   
1  Sunrisers Hyderabad  Royal Challengers Bangalore     1     2  DA Warner   
2  Sunrisers Hyderabad  Royal Challengers Bangalore     1     3  DA Warner   

   ... bye_runs legbye_runs  noball_runs  penalty_runs  batsman_runs  \
0  ...        0           0            0             0             0   
1  ...        0           0            0             0             0   
2  ...        0           0            0             0             4   

   extra_runs  total_runs_y  player_dismissed  dismissal_kind  fielder  
0           0             0               NaN             NaN      NaN  
1           0             0               NaN             NaN      NaN  
2           0             4               NaN             NaN      NaN  

[3 rows x 24 columns]

In [95]:
delivery_df=delivery_df[delivery_df['inning']==2]

In [96]:
delivery_df.shape

(72413, 24)

In [97]:
delivery_df

match_id       city               winner  total_runs_x  inning  \
125            1  Hyderabad  Sunrisers Hyderabad           207       2   
126            1  Hyderabad  Sunrisers Hyderabad           207       2   
127            1  Hyderabad  Sunrisers Hyderabad           207       2   
128            1  Hyderabad  Sunrisers Hyderabad           207       2   
129            1  Hyderabad  Sunrisers Hyderabad           207       2   
...          ...        ...                  ...           ...     ...   
149573     11415  Hyderabad       Mumbai Indians           152       2   
149574     11415  Hyderabad       Mumbai Indians           152       2   
149575     11415  Hyderabad       Mumbai Indians           152       2   
149576     11415  Hyderabad       Mumbai Indians           152       2   
149577     11415  Hyderabad       Mumbai Indians           152       2   

                       batting_team         bowling_team  over  ball  \
125     Royal Challengers Bangalore  Sunrisers Hyderabad     1     1   
126     Royal Challengers Bangalore  Sunrisers Hyderabad     1     2   
127     Royal Challengers Bangalore  Sunrisers Hyderabad     1     3   
128     Royal Challengers Bangalore  Sunrisers Hyderabad     1     4   
129     Royal Challengers Bangalore  Sunrisers Hyderabad     1     5   
...                             ...                  ...   ...   ...   
149573          Chennai Super Kings       Mumbai Indians    20     2   
149574          Chennai Super Kings       Mumbai Indians    20     3   
149575          Chennai Super Kings       Mumbai Indians    20     4   
149576          Chennai Super Kings       Mumbai Indians    20     5   
149577          Chennai Super Kings       Mumbai Indians    20     6   

              batsman  ... bye_runs legbye_runs  noball_runs  penalty_runs  \
125          CH Gayle  ...        0           0            0             0   
126     Mandeep Singh  ...        0           0            0             0   
127     Mandeep Singh  ...        0           0            0             0   
128     Mandeep Singh  ...        0           0            0             0   
129     Mandeep Singh  ...        0           0            0             0   
...               ...  ...      ...         ...          ...           ...   
149573      RA Jadeja  ...        0           0            0             0   
149574      SR Watson  ...        0           0            0             0   
149575      SR Watson  ...        0           0            0             0   
149576      SN Thakur  ...        0           0            0             0   
149577      SN Thakur  ...        0           0            0             0   

        batsman_runs  extra_runs  total_runs_y  player_dismissed  \
125                1           0             1               NaN   
126                0           0             0               NaN   
127                0           0             0               NaN   
128                2           0             2               NaN   
129                4           0             4               NaN   
...              ...         ...           ...               ...   
149573             1           0             1               NaN   
149574             2           0             2               NaN   
149575             1           0             1         SR Watson   
149576             2           0             2               NaN   
149577             0           0             0         SN Thakur   

        dismissal_kind    fielder  
125                NaN        NaN  
126                NaN        NaN  
127                NaN        NaN  
128                NaN        NaN  
129                NaN        NaN  
...                ...        ...  
149573             NaN        NaN  
149574             NaN        NaN  
149575         run out  KH Pandya  
149576             NaN        NaN  
149577             lbw        NaN  

[72413 rows x 24 columns]

In [98]:
delivery_df['current_score']= delivery_df.groupby('match_id').cumsum()['total_runs_y']

In [99]:
delivery_df

match_id       city               winner  total_runs_x  inning  \
125            1  Hyderabad  Sunrisers Hyderabad           207       2   
126            1  Hyderabad  Sunrisers Hyderabad           207       2   
127            1  Hyderabad  Sunrisers Hyderabad           207       2   
128            1  Hyderabad  Sunrisers Hyderabad           207       2   
129            1  Hyderabad  Sunrisers Hyderabad           207       2   
...          ...        ...                  ...           ...     ...   
149573     11415  Hyderabad       Mumbai Indians           152       2   
149574     11415  Hyderabad       Mumbai Indians           152       2   
149575     11415  Hyderabad       Mumbai Indians           152       2   
149576     11415  Hyderabad       Mumbai Indians           152       2   
149577     11415  Hyderabad       Mumbai Indians           152       2   

                       batting_team         bowling_team  over  ball  \
125     Royal Challengers Bangalore  Sunrisers Hyderabad     1     1   
126     Royal Challengers Bangalore  Sunrisers Hyderabad     1     2   
127     Royal Challengers Bangalore  Sunrisers Hyderabad     1     3   
128     Royal Challengers Bangalore  Sunrisers Hyderabad     1     4   
129     Royal Challengers Bangalore  Sunrisers Hyderabad     1     5   
...                             ...                  ...   ...   ...   
149573          Chennai Super Kings       Mumbai Indians    20     2   
149574          Chennai Super Kings       Mumbai Indians    20     3   
149575          Chennai Super Kings       Mumbai Indians    20     4   
149576          Chennai Super Kings       Mumbai Indians    20     5   
149577          Chennai Super Kings       Mumbai Indians    20     6   

              batsman  ... legbye_runs noball_runs  penalty_runs  \
125          CH Gayle  ...           0           0             0   
126     Mandeep Singh  ...           0           0             0   
127     Mandeep Singh  ...           0           0             0   
128     Mandeep Singh  ...           0           0             0   
129     Mandeep Singh  ...           0           0             0   
...               ...  ...         ...         ...           ...   
149573      RA Jadeja  ...           0           0             0   
149574      SR Watson  ...           0           0             0   
149575      SR Watson  ...           0           0             0   
149576      SN Thakur  ...           0           0             0   
149577      SN Thakur  ...           0           0             0   

        batsman_runs  extra_runs  total_runs_y  player_dismissed  \
125                1           0             1               NaN   
126                0           0             0               NaN   
127                0           0             0               NaN   
128                2           0             2               NaN   
129                4           0             4               NaN   
...              ...         ...           ...               ...   
149573             1           0             1               NaN   
149574             2           0             2               NaN   
149575             1           0             1         SR Watson   
149576             2           0             2               NaN   
149577             0           0             0         SN Thakur   

        dismissal_kind    fielder  current_score  
125                NaN        NaN              1  
126                NaN        NaN              1  
127                NaN        NaN              1  
128                NaN        NaN              3  
129                NaN        NaN              7  
...                ...        ...            ...  
149573             NaN        NaN            152  
149574             NaN        NaN            154  
149575         run out  KH Pandya            155  
149576             NaN        NaN            157  
149577             lbw        NaN            157  

[72413 

In [100]:
delivery_df['runs_left']= delivery_df['total_runs_x']+1-delivery_df['current_score']

In [101]:
delivery_df.head(5)

match_id       city               winner  total_runs_x  inning  \
125         1  Hyderabad  Sunrisers Hyderabad           207       2   
126         1  Hyderabad  Sunrisers Hyderabad           207       2   
127         1  Hyderabad  Sunrisers Hyderabad           207       2   
128         1  Hyderabad  Sunrisers Hyderabad           207       2   
129         1  Hyderabad  Sunrisers Hyderabad           207       2   

                    batting_team         bowling_team  over  ball  \
125  Royal Challengers Bangalore  Sunrisers Hyderabad     1     1   
126  Royal Challengers Bangalore  Sunrisers Hyderabad     1     2   
127  Royal Challengers Bangalore  Sunrisers Hyderabad     1     3   
128  Royal Challengers Bangalore  Sunrisers Hyderabad     1     4   
129  Royal Challengers Bangalore  Sunrisers Hyderabad     1     5   

           batsman  ... noball_runs penalty_runs  batsman_runs  extra_runs  \
125       CH Gayle  ...           0            0             1           0   
126  Mandeep Singh  ...           0            0             0           0   
127  Mandeep Singh  ...           0            0             0           0   
128  Mandeep Singh  ...           0            0             2           0   
129  Mandeep Singh  ...           0            0             4           0   

     total_runs_y  player_dismissed  dismissal_kind  fielder  current_score  \
125             1               NaN             NaN      NaN              1   
126             0               NaN             NaN      NaN              1   
127             0               NaN             NaN      NaN              1   
128             2               NaN             NaN      NaN              3   
129             4               NaN             NaN      NaN              7   

     runs_left  
125        207  
126        207  
127        207  
128        205  
129        201  

[5 rows x 26 columns]

In [102]:
delivery_df['balls_left']= 126-(delivery_df['over']*6+delivery_df['ball'])

In [103]:
delivery_df.head(5)

match_id       city               winner  total_runs_x  inning  \
125         1  Hyderabad  Sunrisers Hyderabad           207       2   
126         1  Hyderabad  Sunrisers Hyderabad           207       2   
127         1  Hyderabad  Sunrisers Hyderabad           207       2   
128         1  Hyderabad  Sunrisers Hyderabad           207       2   
129         1  Hyderabad  Sunrisers Hyderabad           207       2   

                    batting_team         bowling_team  over  ball  \
125  Royal Challengers Bangalore  Sunrisers Hyderabad     1     1   
126  Royal Challengers Bangalore  Sunrisers Hyderabad     1     2   
127  Royal Challengers Bangalore  Sunrisers Hyderabad     1     3   
128  Royal Challengers Bangalore  Sunrisers Hyderabad     1     4   
129  Royal Challengers Bangalore  Sunrisers Hyderabad     1     5   

           batsman  ... penalty_runs batsman_runs  extra_runs  total_runs_y  \
125       CH Gayle  ...            0            1           0             1   
126  Mandeep Singh  ...            0            0           0             0   
127  Mandeep Singh  ...            0            0           0             0   
128  Mandeep Singh  ...            0            2           0             2   
129  Mandeep Singh  ...            0            4           0             4   

     player_dismissed  dismissal_kind  fielder  current_score  runs_left  \
125               NaN             NaN      NaN              1        207   
126               NaN             NaN      NaN              1        207   
127               NaN             NaN      NaN              1        207   
128               NaN             NaN      NaN              3        205   
129               NaN             NaN      NaN              7        201   

     balls_left  
125         119  
126         118  
127         117  
128         116  
129         115  

[5 rows x 27 columns]

In [104]:
delivery_df.isnull().sum()

match_id                0
city                  832
winner                 27
total_runs_x            0
inning                  0
batting_team            0
bowling_team            0
over                    0
ball                    0
batsman                 0
non_striker             0
bowler                  0
is_super_over           0
wide_runs               0
bye_runs                0
legbye_runs             0
noball_runs             0
penalty_runs            0
batsman_runs            0
extra_runs              0
total_runs_y            0
player_dismissed    68860
dismissal_kind      68860
fielder             69855
current_score           0
runs_left               0
balls_left              0
dtype: int64

In [105]:
delivery_df['player_dismissed']=delivery_df['player_dismissed'].fillna('0')

In [106]:
delivery_df['player_dismissed']=delivery_df['player_dismissed'].apply(lambda x:x if x=='0' else '1') 

In [107]:
delivery_df['player_dismissed']= delivery_df['player_dismissed'].astype('int')

In [108]:
 wickets=delivery_df.groupby(['match_id']).cumsum()['player_dismissed'].values

In [109]:
delivery_df['wickets']= 10-wickets

In [110]:
delivery_df.head(5)

match_id       city               winner  total_runs_x  inning  \
125         1  Hyderabad  Sunrisers Hyderabad           207       2   
126         1  Hyderabad  Sunrisers Hyderabad           207       2   
127         1  Hyderabad  Sunrisers Hyderabad           207       2   
128         1  Hyderabad  Sunrisers Hyderabad           207       2   
129         1  Hyderabad  Sunrisers Hyderabad           207       2   

                    batting_team         bowling_team  over  ball  \
125  Royal Challengers Bangalore  Sunrisers Hyderabad     1     1   
126  Royal Challengers Bangalore  Sunrisers Hyderabad     1     2   
127  Royal Challengers Bangalore  Sunrisers Hyderabad     1     3   
128  Royal Challengers Bangalore  Sunrisers Hyderabad     1     4   
129  Royal Challengers Bangalore  Sunrisers Hyderabad     1     5   

           batsman  ... batsman_runs extra_runs  total_runs_y  \
125       CH Gayle  ...            1          0             1   
126  Mandeep Singh  ...            0          0             0   
127  Mandeep Singh  ...            0          0             0   
128  Mandeep Singh  ...            2          0             2   
129  Mandeep Singh  ...            4          0             4   

     player_dismissed  dismissal_kind  fielder  current_score  runs_left  \
125                 0             NaN      NaN              1        207   
126                 0             NaN      NaN              1        207   
127                 0             NaN      NaN              1        207   
128                 0             NaN      NaN              3        205   
129                 0             NaN      NaN              7        201   

     balls_left  wickets  
125         119       10  
126         118       10  
127         117       10  
128         116       10  
129         115       10  

[5 rows x 28 columns]

In [111]:
delivery_df['crr']=delivery_df['current_score']*6/(120-delivery_df['balls_left'])

In [112]:
delivery_df.head()

match_id       city               winner  total_runs_x  inning  \
125         1  Hyderabad  Sunrisers Hyderabad           207       2   
126         1  Hyderabad  Sunrisers Hyderabad           207       2   
127         1  Hyderabad  Sunrisers Hyderabad           207       2   
128         1  Hyderabad  Sunrisers Hyderabad           207       2   
129         1  Hyderabad  Sunrisers Hyderabad           207       2   

                    batting_team         bowling_team  over  ball  \
125  Royal Challengers Bangalore  Sunrisers Hyderabad     1     1   
126  Royal Challengers Bangalore  Sunrisers Hyderabad     1     2   
127  Royal Challengers Bangalore  Sunrisers Hyderabad     1     3   
128  Royal Challengers Bangalore  Sunrisers Hyderabad     1     4   
129  Royal Challengers Bangalore  Sunrisers Hyderabad     1     5   

           batsman  ... extra_runs total_runs_y  player_dismissed  \
125       CH Gayle  ...          0            1                 0   
126  Mandeep Singh  ...          0            0                 0   
127  Mandeep Singh  ...          0            0                 0   
128  Mandeep Singh  ...          0            2                 0   
129  Mandeep Singh  ...          0            4                 0   

     dismissal_kind  fielder  current_score  runs_left  balls_left  wickets  \
125             NaN      NaN              1        207         119       10   
126             NaN      NaN              1        207         118       10   
127             NaN      NaN              1        207         117       10   
128             NaN      NaN              3        205         116       10   
129             NaN      NaN              7        201         115       10   

     crr  
125  6.0  
126  3.0  
127  2.0  
128  4.5  
129  8.4  

[5 rows x 29 columns]

In [113]:
delivery_df['rrr']=(delivery_df['runs_left']*6)/(delivery_df['balls_left'])

In [114]:
delivery_df.head()

match_id       city               winner  total_runs_x  inning  \
125         1  Hyderabad  Sunrisers Hyderabad           207       2   
126         1  Hyderabad  Sunrisers Hyderabad           207       2   
127         1  Hyderabad  Sunrisers Hyderabad           207       2   
128         1  Hyderabad  Sunrisers Hyderabad           207       2   
129         1  Hyderabad  Sunrisers Hyderabad           207       2   

                    batting_team         bowling_team  over  ball  \
125  Royal Challengers Bangalore  Sunrisers Hyderabad     1     1   
126  Royal Challengers Bangalore  Sunrisers Hyderabad     1     2   
127  Royal Challengers Bangalore  Sunrisers Hyderabad     1     3   
128  Royal Challengers Bangalore  Sunrisers Hyderabad     1     4   
129  Royal Challengers Bangalore  Sunrisers Hyderabad     1     5   

           batsman  ... total_runs_y player_dismissed  dismissal_kind  \
125       CH Gayle  ...            1                0             NaN   
126  Mandeep Singh  ...            0                0             NaN   
127  Mandeep Singh  ...            0                0             NaN   
128  Mandeep Singh  ...            2                0             NaN   
129  Mandeep Singh  ...            4                0             NaN   

     fielder  current_score  runs_left  balls_left  wickets  crr        rrr  
125      NaN              1        207         119       10  6.0  10.436975  
126      NaN              1        207         118       10  3.0  10.525424  
127      NaN              1        207         117       10  2.0  10.615385  
128      NaN              3        205         116       10  4.5  10.603448  
129      NaN              7        201         115       10  8.4  10.486957  

[5 rows x 30 columns]

In [115]:
def result(row):
    return 1 if row['batting_team']== row['winner'] else 0

In [116]:
delivery_df['result']=delivery_df.apply(result,axis=1)

In [117]:
delivery_df.sample(5)

match_id           city                       winner  total_runs_x  \
58058        306        Chennai  Royal Challengers Bangalore           185   
64747        343        Chennai              Kings XI Punjab           156   
60163        317      Bangalore        Kolkata Knight Riders           165   
18226        109          Delhi             Delhi Daredevils           176   
149022     11413  Visakhapatnam               Delhi Capitals           171   

        inning                 batting_team                 bowling_team  \
58058        2               Mumbai Indians  Royal Challengers Bangalore   
64747        2          Chennai Super Kings              Kings XI Punjab   
60163        2  Royal Challengers Bangalore        Kolkata Knight Riders   
18226        2             Delhi Daredevils               Mumbai Indians   
149022       2               Delhi Capitals          Sunrisers Hyderabad   

        over  ball     batsman  ... player_dismissed dismissal_kind  fielder  \
58058      7     2   RG Sharma  ...                0            NaN      NaN   
64747     15     2    DJ Bravo  ...                0            NaN      NaN   
60163     12     1  DL Vettori  ...                0            NaN      NaN   
18226      8     3    S Dhawan  ...                0            NaN      NaN   
149022     9     6     SS Iyer  ...                0            NaN      NaN   

        current_score  runs_left  balls_left  wickets       crr        rrr  \
58058              53        133          82        8  8.368421   9.731707   
64747              96         61          34        5  6.697674  10.764706   
60163              57        109          53        5  5.104478  12.339623   
18226              70        107          75        8  9.333333   8.560000   
149022             78         94          66       10  8.666667   8.545455   

        result  
58058        0  
64747        0  
60163        0  
18226        1  
149022       1  

[5 rows x 31 columns]

In [118]:
final_df=delivery_df[['batting_team','bowling_team','city','runs_left','balls_left','wickets','total_runs_x','crr','rrr','result']]

In [119]:
final_df=final_df.sample(final_df.shape[0])

In [120]:
final_df.sample()

batting_team                 bowling_team    city  runs_left  \
938  Kings XI Punjab  Royal Challengers Bangalore  Indore         49   

     balls_left  wickets  total_runs_x    crr   rrr  result  
938          56        8           148  9.375  5.25       1

In [121]:
final_df.isnull().sum()

batting_team      0
bowling_team      0
city            832
runs_left         0
balls_left        0
wickets           0
total_runs_x      0
crr               0
rrr               5
result            0
dtype: int64

In [122]:
final_df.dropna(inplace=True)

In [123]:
final_df.isnull().sum()

batting_team    0
bowling_team    0
city            0
runs_left       0
balls_left      0
wickets         0
total_runs_x    0
crr             0
rrr             0
result          0
dtype: int64

In [124]:
final_df = final_df[final_df['balls_left'] != 0]

In [125]:
final_df.head(3)

batting_team                 bowling_team          city  runs_left  \
25493  Delhi Daredevils          Chennai Super Kings  Johannesburg         93   
29102    Mumbai Indians              Kings XI Punjab     Centurion         70   
17064  Delhi Daredevils  Royal Challengers Bangalore     Bangalore         33   

       balls_left  wickets  total_runs_x       crr       rrr  result  
25493          63        7           163  7.473684  8.857143       0  
29102          73        8           119  6.382979  5.753425       1  
17064          37        7           154  8.819277  5.351351       1

In [126]:
final_df['balls_left'].unique()

array([ 63,  73,  37,  75,  70, 104,  91,  38,  34,  81,  45,  31, 107,
        72,  96,  87,  65,  39,  95, 101,  60, 102,  51,  16,  97,  80,
       112, 118,  15,  22,  49,  28,  11,  40, 103,  62, 109,  56,  48,
        68,  21,  82,  29,  13,  33,  92,  93,  76,  20,  90,  27,  77,
         9,   6,  89, 113,  12, 116,  78,  30,  59,  53,  14,   5,  52,
        57, 110, 119,  64,  98,   8,  46,  67,  32,  86,  18,  50,  84,
        58, 100, 114,  69,  83, 117, 115,  66,  55,  74,   7,  85,  35,
        88,  19, 105,  71,  10,  42,  94,  54,  47,  43,  24, 111,  41,
        17,   4,  36,  61, 106,  79,  26,   3,  99,  44,   2,  23, 108,
        25,   1,  -1,  -2], dtype=int64)

In [127]:
X = final_df.iloc[:,:-1]
y = final_df.iloc[:,-1]
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=1)

In [128]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder

trf = ColumnTransformer([
    ('trf',OneHotEncoder(sparse=False,drop='first'),['batting_team','bowling_team','city'])
]
,remainder='passthrough')

In [129]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.pipeline import Pipeline

In [130]:
pipe = Pipeline(steps=[
    ('step1',trf),
    ('step2',LogisticRegression(solver='liblinear'))
])

In [131]:
pipe.fit(X_train,y_train)

Pipeline(steps=[('step1',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('trf',
                                                  OneHotEncoder(drop='first',
                                                                sparse=False),
                                                  ['batting_team',
                                                   'bowling_team', 'city'])])),
                ('step2', LogisticRegression(solver='liblinear'))])

In [133]:
y_pred = pipe.predict(X_test)

In [134]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test,y_pred)

0.8033499194057047

In [135]:
#pipe = Pipeline(steps=[
 #   ('step1',trf),
  #  ('step2',RandomForestClassifier())
#])

In [136]:
#pipe.fit(X_train,y_train)

In [137]:
#y_pred = pipe.predict(x_test)

In [138]:
#accuracy_score(y_test,y_pred)

In [139]:
pipe.predict_proba(X_test)[8
                          ]

array([0.48168594, 0.51831406])

In [140]:
def match_progression(x_df,match_id,pipe):
    match = x_df[x_df['match_id'] == match_id]
    match = match[(match['ball'] == 6)]
    temp_df = match[['batting_team','bowling_team','city','runs_left','balls_left','wickets','total_runs_x','crr','rrr']].dropna()
    temp_df = temp_df[temp_df['balls_left'] != 0]
    result = pipe.predict_proba(temp_df)
    temp_df['lose'] = np.round(result.T[0]*100,1)
    temp_df['win'] = np.round(result.T[1]*100,1)
    temp_df['end_of_over'] = range(1,temp_df.shape[0]+1)
    
    target = temp_df['total_runs_x'].values[0]
    runs = list(temp_df['runs_left'].values)
    new_runs = runs[:]
    runs.insert(0,target)
    temp_df['runs_after_over'] = np.array(runs)[:-1] - np.array(new_runs)
    wickets = list(temp_df['wickets'].values)
    new_wickets = wickets[:]
    new_wickets.insert(0,10)
    wickets.append(0)
    w = np.array(wickets)
    nw = np.array(new_wickets)
    temp_df['wickets_in_over'] = (nw - w)[0:temp_df.shape[0]]
    
    print("Target-",target)
    temp_df = temp_df[['end_of_over','runs_after_over','wickets_in_over','lose','win']]
    return temp_df,target

In [141]:
temp_df,target = match_progression(delivery_df,74,pipe)
temp_df

Target- 178


end_of_over  runs_after_over  wickets_in_over  lose   win
10459            1                3                0  58.0  42.0
10467            2                8                0  52.8  47.2
10473            3                1                0  59.4  40.6
10479            4                7                1  70.8  29.2
10485            5               12                0  60.9  39.1
10491            6               13                0  48.7  51.3
10497            7                9                0  42.8  57.2
10505            8               15                0  28.7  71.3
10511            9                7                0  26.5  73.5
10518           10               17                0  14.6  85.4
10524           11                9                1  20.1  79.9
10530           12                9                0  16.6  83.4
10536           13                8                0  14.3  85.7
10542           14                8                0  12.2  87.8
10548           15                5                1  21.2  78.8
10555           16                8                1  30.0  70.0
10561           17                8                2  56.3  43.7
10567           18                6                1  71.4  28.6
10573           19                8                2  90.0  10.0

In [142]:
teams

['Sunrisers Hyderabad',
 'Mumbai Indians',
 'Royal Challengers Bangalore',
 'Kolkata Knight Riders',
 'Kings XI Punjab',
 'Chennai Super Kings',
 'Rajasthan Royals',
 'Delhi Capitals']

In [143]:
delivery_df['city'].unique()

array(['Hyderabad', 'Bangalore', 'Mumbai', 'Indore', 'Kolkata', 'Delhi',
       'Chandigarh', 'Jaipur', 'Chennai', 'Cape Town', 'Port Elizabeth',
       'Durban', 'Centurion', 'East London', 'Johannesburg', 'Kimberley',
       'Bloemfontein', 'Ahmedabad', 'Cuttack', 'Nagpur', 'Dharamsala',
       'Visakhapatnam', 'Pune', 'Raipur', 'Ranchi', 'Abu Dhabi',
       'Sharjah', nan, 'Mohali', 'Bengaluru'], dtype=object)

In [144]:
import pickle
pickle.dump(pipe,open('pipe.pkl','wb'))